File: go/linkedlist/circular_linkedlist/circular_linkedlist.go
package circular_linkedlist

import "fmt"

//循环链表实现

//链表节点定义
type ListNode struct {
	val  interface{}
	next *ListNode
}

//新建节点
func newNode(val interface{}) *ListNode {
	return &ListNode{val, nil}
}

//链表定义
type LinkedList struct {
	head *ListNode
	tail *ListNode
	size int
}

func New() *LinkedList {
	return &LinkedList{nil, nil, 0}
}

func (lis *LinkedList) Find(val interface{}) *ListNode {
	p := lis.head
	for i := 0; i < lis.size && p.val != val; i++ {
		p = p.next
	}
	return p
}

//添加节点至链表末尾
func (lis *LinkedList) PushBack(val interface{}) *ListNode {
	node := newNode(val)
	if lis.head == nil {
		lis.head = node
	} else {
		node.next = lis.head
		lis.tail.next = node
	}

	lis.tail = node
	lis.size++
	return node
}

//在某个节点之后插入数据
func (lis *LinkedList) PushAfter(p *ListNode, val interface{}) *ListNode {
	if p == nil {
		return nil
	}

	node := newNode(val)
	node.next = p.next
	p.next = node
	if p == lis.tail {
		lis.tail = node
	}
	lis.size++
	return node
}

//在某个节点之前插入数据
func (lis *LinkedList) PushBefore(p *ListNode, val interface{}) *ListNode {
	if p == nil || lis.head == nil {
		return nil
	}

	if p == lis.head {
		return lis.PushBack(val)
	}

	prev := lis.head
	for i := 0; i < lis.size && prev.next != p; i++ {
		prev = prev.next
	}

	return lis.PushAfter(prev, val)
}

func (lis *LinkedList) Size() int {
	return lis.size
}

//删除节点
func (lis *LinkedList) Delete(p *ListNode) bool {
	if p == nil || lis.head == nil {
		return false
	}

	if p == lis.head {
		lis.head = p.next
	} else {
		prev := lis.head
		for i := 0; i < lis.size && prev.next != p; i++ {
			prev = prev.next
		}
		prev.next = p.next
	}
	lis.size--
	return true
}

//根据值删除节点
func (lis *LinkedList) DeleteVal(val interface{}) bool {
	return lis.Delete(lis.Find(val))
}

//打印所有节点数据
func (lis *LinkedList) PrintData() {
	p := lis.head
	for i := 0; i < lis.size; i++ {
		fmt.Print(p.val, " ")
		p = p.next
	}
	fmt.Println()
}

File: python/linkedlist/circular_linked_list.py

# 循环链表实现

class ListNode:

    def __init__(self, val):
        self.val = val
        self.prev = None
        self.next = None


class LinkedList:

    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    # 添加元素至链表尾部
    def add(self, val: int):
        node = ListNode(val)
        if not self.head:
            self.head = node
        if self.tail:
            self.tail.next = node

        self.tail = node
        self.tail.next = self.head
        self.size += 1

    # 在某个节点之后添加节点
    def addAfter(self, p: ListNode, val: int):
        if not p:
            return
        node = ListNode(val)
        node.next = p.next
        p.next = node

        if p == self.tail:
            self.tail = node
            self.tail.next = self.head
        self.size += 1

    # 根据值删除节点
    def delete(self, val: int):
        node = self.find(val)
        if node is None:
            return

        if node == self.head:
            self.head = self.head.next
        else:
            prev = lis.head
            for i in range(self.size):
                if prev.next == node:
                    break
            prev.next = node.next.next
        self.size -= 1

    # 根据元素值查找节点
    def find(self, val: int):
        p = self.head
        for i in range(self.size):
            if p.val == val:
                return p
            p = p.next
        return p

    # 打印链表中所有元素
    def print(self):
        p = self.head
        for i in range(self.size):
            print(p.val, end=' ')
            p = p.next
        print()

    def __size__(self):
        return self.size


if __name__ == '__main__':
    lis = LinkedList()
    lis.add(10)

    lis.delete(10)
    lis.print()

    lis.add(8)
    lis.add(21)
    lis.add(3)
    lis.add(9)

    print(lis.__size__())

    lis.print()

    n = lis.find(10)
    print(n.val)

    lis.addAfter(n, 15)
    lis.print()

    n9 = lis.find(9)
    lis.addAfter(n9, 24)
    lis.print()

    lis.delete(24)
    lis.print()
    print(lis.size)
